packages/protoplugin/src/ecmascript/transpile.ts

```diff
@@ -0,0 +1,155 @@
+// Copyright 2021-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { FileInfo } from "./generated-file.js";
+import ts from "typescript";
+import {
+  createDefaultMapFromNodeModules,
+  createSystem,
+  createVirtualCompilerHost,
+} from "@typescript/vfs";
+
+/* eslint-disable import/no-named-as-default-member */
+
+// The default options used to auto-transpile if needed.
+const defaultOptions: ts.CompilerOptions = {
+  // Type checking
+  strict: false,
+
+  // modules
+  module: ts.ModuleKind.ES2020,
+  moduleResolution: ts.ModuleResolutionKind.NodeJs,
+  noResolve: true,
+  resolveJsonModule: false,
+
+  // emit
+  emitBOM: false,
+  importsNotUsedAsValues: ts.ImportsNotUsedAsValues.Preserve,
+  newLine: ts.NewLineKind.LineFeed,
+  preserveValueImports: false,
+
+  // JavaScript Support
+  checkJs: false,
+
+  // Language and Environment
+  lib: [],
+  moduleDetection: "force",
+  target: ts.ScriptTarget.ES2017,
+
+  // Completeness
+  skipLibCheck: true,
+  skipDefaultLibCheck: false,
+};
+
+/**
+ * Create a transpiler using the given compiler options, which will compile the
+ * content provided in the files array.
+ *
+ * Note:  this library intentionally transpiles with a pinned older version of
+ * TypeScript for stability.  This version is denoted in this workspace's
+ * package.json.  For the default set of compiler options, we use a lenient
+ * set of options because the general goal is to emit code as best as we can.
+ * For a list of the options used, see `defaultOptions` above.
+ *
+ * If this is not desirable for plugin authors, they are free to provide their
+ * own transpile function as part of the plugin initialization.  If one is
+ * provided, it will be invoked instead and the framework's auto-transpilation
+ * will be bypassed.
+ *
+ * In addition, note that there is a dependency on @typescript/vfs in the
+ * top-level package as well as this package.  This is to avoid npm hoisting
+ * @typescript/vfs to the top-level node_modules directory, which then causes
+ * type mismatches when trying to use it with this package's version of
+ * TypeScript.  Ideally we would use something like Yarn's nohoist here, but
+ * npm does not support that yet.
+ */
+function createTranspiler(options: ts.CompilerOptions, files: FileInfo[]) {
+  const fsMap = createDefaultMapFromNodeModules({
+    target: options.target,
+  });
```

Comment:
Looking at the [source of this function](https://github.com/microsoft/TypeScript-Website/blob/v2/packages/typescript-vfs/src/index.ts#L194), I think we should be able to replace this with just `const fsMap = new Map<string, string>()`.

The behavior of `createDefaultMapFromNodeModules` is:
> Sets up a Map with lib contents by grabbing the necessary files from the local copy of typescript via the file system.

This behavior is actually undesirable in our case, because we want to avoid hitting the file system. We don't want to introduce any variance in the output depending on the local file system, and we intentionally ignore any resolution errors anyway.

I _think_ `createDefaultMapFromNodeModules` doesn't do anything because we pass an empty `lib` option, so this change will probably not affect the outcome. But not using the function in the first place would remove any doubt, and make the intent clear.

WDYT?

Reply:
So, it's weird.  When I was first coding this, I used the empty `Map` example you have and was getting errors with transpilation.  After reading the docs, it seemed I needed to add this `createDefaultMap...` call.  Once I did that, everything started working.

But, when I take it out now, things still seem to be working, so I'm not sure if something I did also fixed that issue.  All tests are passing, so I'm guessing it's OK?  I'll remove and we can revisit if something blows up.

Reply:
The tests are pretty lightweight, and don't use any imports or lib types. But I think this is fine as is for now. 

Once we add documentation for this feature, we can provide a simple example (maybe a simple Twirp client generator using fetch - it would use both imports and lib types then), and gain confidence that it works as expected.